### PR TITLE
SorrIssue328 Convert Type Hints gDoc to Markdown

### PR DESCRIPTION
**Pull request by hhxjqm**
Instructions told me that "The entire conversion flow is in dev_scripts/convert_gdoc_to_markdown.sh", but I can't find it, now I know it called ./dev_scripts/convert_docx_to_markdown.sh.

I redo it and here is update

### CODE REVIEW

**Comment by DanilYachmenev** [comment]
@hhxjqm pls carefully read https://github.com/sorrentum/sorrentum/blob/master/docs/First_review_process.md#save-reviewer-time 

treat your PR accordingly, use labels and assign team leaders first

**Comment by DanilYachmenev** [comment]
also, should we consider #334 obsolete?

File: docs/Python_Type_hints.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+## Invoke tasks
+
+For some reason `invoke` does not like type hints, so we
+
+- Omit type hints for `invoke` tasks, i.e. functions with the `@task` decorator
+- Put `# type: ignore` so that `mypy` does not complain
+
+Example:
+  ```python
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:59]: nicee

File: docs/Python_Type_hints.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+## Annotation for `kwargs`
+
+- We use `kwargs: Any` and not `kwargs: Dict[str, Any]`
+- `*` always binds to a `Tuple`, and `**` always binds to a `Dict[str, Any]`.
+  Because of this restriction, type hints only need you to define the types of
+  the contained arguments. The type checker automatically adds the
+  `Tuple[_, ...]` and `Dict[str, _]` container types.
+- [Reference article](https://adamj.eu/tech/2021/05/11/python-type-hints-args-and-kwargs/)
+## `Any`
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:75]: maintain empty spaces before and after headers

File: docs/Python_Type_hints.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  1. Type hint when assigning the model parameter in ctor:
+     ```bash
+     self._model: Optional[sklearn.base.BaseEstimator] = None
+     ```
+  2. Cast a type to the model parameter after asserting that it is not `None`:
+     ```bash
+     hdbg.dassert_is_not(self._model, None) self._model = cast(sklearn.base.BaseEstimator, self._model)
+     ```
+
+# Disabling `mypy` errors
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:159]: This should be a header format, right?
for some reason displayed as a comment

Comment (hhxjqm) [reply docs/Python_Type_hints.md:161]: It works well in display the rich diff. Still trying figure out.

Comment (DanilYachmenev) [reply docs/Python_Type_hints.md:161]: Well, it looks correct indeed and also reflected in TOH
So IMO its ok

**Review by DanilYachmenev — COMMENTED** [review]
lgtm!

**Comment by DanilYachmenev** [comment]
@samarth9008 could you ptal, check for some omissions that I could make and check out the unresolved comment?
If everything is OK, passing to GP

File: docs/Python_Type_hints.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  - Omit type hints for `invoke` tasks, i.e. functions with the `@task` decorator
+  - Put `# type: ignore` so that `mypy` does not complain
+
+- Example:
+  ```python
+  @task def run_qa_tests( # type: ignore 
+    ctx, 
+    stage="dev", 
+    version="", 
+  ):
```
Comment (samarth9008) [inline docs/Python_Type_hints.md:65]:
```python
@task 
def run_qa_tests( # type: ignore 
  ctx, 
  stage="dev", 
  version="", 
):
```

File: docs/Python_Type_hints.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+- We try to avoid it everywhere when possible
+
+## `np.array` and `np.ndarray`
+
+- If you get something like the following lint:
+  ```bash
+  dataflow/core/nodes/sklearn_models.py:537:[amp_mypy] error: Function "numpy.core.multiarray.array" is not valid as a type [valid-type]  
+  ```
+- Then the problem is probably that a parameter that the lint is related to
+  has been types as `np.array` while it should be typed as `np.ndarray`:
```
Comment (samarth9008) [inline docs/Python_Type_hints.md:89]: I think it should be **has been typed as**

File: docs/Python_Type_hints.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  the following lint appears:
+  ```bash
+  dataflow/core/nodes/sklearn_models.py:155:[amp_mypy] error: "None" has no attribute "fit"
+  ```
+- A solution is to
+  - 1. Type hint when assigning the model parameter in ctor:
+     ```bash
+     self._model: Optional[sklearn.base.BaseEstimator] = None
+     ```
+  - 2. Cast a type to the model parameter after asserting that it is not `None`:
```
Comment (samarth9008) [inline docs/Python_Type_hints.md:156]: Is it okay to have number as well as bullet points? Wdyt @DanilYachmenev

Comment (DanilYachmenev) [reply docs/Python_Type_hints.md:156]: hmm, looked ok at first but now I think that there is no order to the points so lets remove the numbers

File: docs/Python_Type_hints.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+     ```bash
+     hdbg.dassert_is_not(self._model, None) self._model = cast(sklearn.base.BaseEstimator, self._model)
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:158]: also split on 2 lines
```
hdbg.dassert_is_not(self._model, None)
self._model = cast(sklearn.base.BaseEstimator, self._model)
```

File: docs/Python_Type_hints.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  ```bash
+  self._model = self._model.fit(...)
+  ```
+  the following lint appears:
+  ```bash
+  dataflow/core/nodes/sklearn_models.py:155:[amp_mypy] error: "None" has no attribute "fit"
+  ```
+- A solution is to
+  - 1. Type hint when assigning the model parameter in ctor:
+     ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:153]: this one is not `bash` but `python`

`bash` is put for the cases when we put commands from bash calls or log outputs in cmd
all the code that is from the actual code base should be backticked with ```python

File: docs/Python_Type_hints.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+  ```
+
+# Handling the annoying `"None" has no attribute`
+
+- In some model classes `self._model` parameter is being assigned to `None` in
+  ctor and being set after calling `set_fit_state` method
+- The problem is that statically it's not possible to understand that someone
+  will call `set_fit_state` before using `self._model`, so when a model's method
+  is applied:
+  ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:144]: python

File: docs/Python_Type_hints.md [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+  dbg.dassert_is_not(var, None)
+  test_func(arg=var)
+  ```
+- Sometimes `mypy` doesn't pick up the `None` check, and warns that the function
+  expects a `bool` rather than an `Optional[bool]`. In that case, the solution
+  is to explicitly use `typing.cast` on the argument when passing it in, note
+  that `typing.cast` has no runtime effects and is purely for type checking.
+- Here're the relevant [docs](https://mypy.readthedocs.io/en/stable/casts.html)
+- So the solution would be:
+  ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:131]: python

File: docs/Python_Type_hints.md [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+## `np.array` and `np.ndarray`
+
+- If you get something like the following lint:
+  ```bash
+  dataflow/core/nodes/sklearn_models.py:537:[amp_mypy] error: Function "numpy.core.multiarray.array" is not valid as a type [valid-type]  
+  ```
+- Then the problem is probably that a parameter that the lint is related to
+  has been types as `np.array` while it should be typed as `np.ndarray`:
+  ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:90]: python

File: docs/Python_Type_hints.md [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+  `x_vals: np.array` -> `x_vals: np.ndarray`
+  ```
+
+# Handling the annoying `Incompatible types in assignment`
+
+- `mypy` assigns a single type to each variable for its entire scope
+- The problem is in common idioms where we use the same variable to store
+  different representations of the same data
+  ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:99]: python

File: docs/Python_Type_hints.md [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+  output : str = ...
+  output = output.split("\n")
+  ...
+  # Process output.
+  ...
+  output = "\n".join(output)
+  ```
+- Unfortunately the proper solution is to use different variables
+  ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:108]: python

File: docs/Python_Type_hints.md [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+  output : str = ...
+  output_as_array = output.split("\n")
+  ...
+  # Process output.
+  ...
+  output = "\n".join(output_as_array)
+  ```
+- Another case could be:
+  ```bash
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:117]: python

File: docs/Python_Type_hints.md (hunk 6 above)
Comment (hhxjqm) [reply docs/Python_Type_hints.md:156]: There are numbers show in Gdoc, if it is not matter, I will remove them.

File: docs/Python_Type_hints.md [hunk 15]
```diff
@@ -0,0 +1,233 @@
+# Python - Type hints
```
Comment (DanilYachmenev) [inline docs/Python_Type_hints.md:1]: Oh, and last thing
lets name it just `Type hints` and the doc name would be `Type_hints.md`

**Review by DanilYachmenev — COMMENTED** [review]
change the name with filename and lgtm

**Comment by samarth9008** [comment]
LGTM

**Comment by hhxjqm** [comment]
@DanilYachmenev It's been two days, do I need to ask for new task?

**Comment by samarth9008** [comment]
> @DanilYachmenev It's been two days, do I need to ask for new task?

We will file more and let you know. Now @gpsaggese will take a look and merge it.

**Review by gpsaggese — COMMENTED** [review]
LG.

@DanilYachmenev and @samarth9008 feel free to strike out the gdocs / move them to OLD as I did for this, so we know what's left
